flux-typeck/src/type_env/paths_tree.rs

```diff
@@ -262,6 +262,21 @@ impl PathsTree {
                             _ => panic!("Unsupported Deref: {elem:?} {ty:?}"),
                         }
                     }
+                    PlaceElem::Index(_) => {
+                        let ty = ptr.borrow().expect_owned();
+                        match ty.kind() {
+                            TyKind::Indexed(BaseTy::Array(arr_ty, _), _)
+                            | TyKind::Exists(BaseTy::Array(arr_ty, _), _) => {
+                                let (rk, ty) =
+                                    Self::lookup_ty(genv, rcx, RefKind::Shr, arr_ty, place_proj)?;
```

Comment:
Ok, this is the main issue. This is technically wrong. Right now `LookupKind` can be either `Ref` or `Node` reflecting the fact that the place can either point to a _path_ in the tree (something you can take a strong reference to) or nested inside a reference (mutable or shared). We need to add an extra case saying the place can also point inside an array/slice. I'd do the following.

First, modify `LookupKind`

```rust
enum LookupKind {
    Strg(Path, NodePtr),
    Weak(WeakKind, Ty),
}
    
enum WeakKind {
    Shr,
    Mut,
    Array,
}
```

Second, modify `lookup_ty` to take a `WeakKind` instead of a `RefKind` (`FoldResult` should be modified accordingly). Note that `WeakKind` doesn't affect how refinement checking works as we only care whether we know the exact location or not (so we know if we can strongly borrow it or not), but we keep track of it to avoid shooting ourselves in the foot.

I don't love the term `Weak`, but I don't have anything better (actually the main reason I haven't implemented this is that I don't have a better name XD)

We should also add tests for taking references into the array:

```rust
fn foo() {
    let mut arr = [1, 2, 3];
    let x = &arr[0];
    let y = &mut arr[0];
}    
```

Reply:
re weak vs strg terminology:

I think using strg vs weak is misleading. I'm just realizing this after writing the paper, but the relevant distinction is whether we know the exact location or not (strong updates are a consequence of knowing the exact location). I which we had some terminology for that.